contracts/carrot-app/src/yield_sources/yield_type.rs

```diff
@@ -0,0 +1,93 @@
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{coins, Coin, CosmosMsg, Deps, Env, SubMsg, Uint128};
+
+use crate::contract::{App, AppResult};
+
+use super::{mars, osmosis_cl_pool};
+
+#[cw_serde]
+pub enum YieldType {
+    /// For osmosis CL Pools, you need a pool id to do your deposit, and that's all
+    ConcentratedLiquidityPool(ConcentratedPoolParams),
+    /// For Mars CL Pools, you just need to deposit in the RedBank
+    /// You need to indicate the denom of the funds you want to deposit
+    Mars(String),
+}
+
+impl YieldType {
+    pub fn deposit(
+        self,
+        deps: Deps,
+        env: &Env,
+        funds: Vec<Coin>,
+        app: &App,
+    ) -> AppResult<Vec<SubMsg>> {
+        match self {
+            YieldType::ConcentratedLiquidityPool(params) => {
+                osmosis_cl_pool::deposit(deps, env, params, funds, app)
+            }
+            YieldType::Mars(denom) => mars::deposit(deps, denom, funds[0].amount, app),
+        }
+    }
+
+    pub fn withdraw(
+        self,
+        deps: Deps,
+        amount: Option<Uint128>,
+        app: &App,
+    ) -> AppResult<Vec<CosmosMsg>> {
```

Comment:
Would be great to have a list of tokens returned there as well, i.e. how many and which tokens the dev can expect to receive after executing this msg. 

The dev can 
- Query all the values of the enabled yield strategies
- Calculate how much to withdraw from each, construct msgs
- Prepare swap messages if needed
- Deposit into other strategy with (potentially swapped) funds. 
And all without the need of replies

Would that be possible to implement?

Reply:
Yes. I have planned a SimulateRebalance query endpoint. I will also do a SimulateDeposit endpoint for those purposes

Reply:
The replies were here to avoid simulations and have the exact swapped funds deposited instead of relying on estimates that may change during execution.

Reply:
We'll have to sync on this, I'm not sure what the idea is behind that

Reply:
> The replies were here to avoid simulations and have the exact swapped funds deposited instead of relying on estimates that may change during execution.

Simulations during execution are accurate as long as you don't do any action on the pool before you use a result of that simulation.

Reply:
That's the thing. As we're doing potentially multiple swaps, we're not able to trust those simulations. We can try if you prefer. You don't like replies ?

Reply:
Yeah makes sense! But in general our rebalance approach should probably be 
1. Withdraw
2. Swap (Optional)
3. Deposit

If we add a simulation we can prevent reply 1 -> 2, independent of wether we need to swap or not

Reply:
Replies just makes reading the code a bit messier imo and requires state caching